### DISCUSSION TITLE
How to interpret signature a(sa{sv})

### DISCUSSION THREAD

**Discussion by Krulle86** [Kistler-Group/sdbus-cpp#445] [Q&A] [answered]
Hello togehter,

i have a issue with the above DBus signature (sdbus version 1.0.0). 
I interpret it as the following data type: `sdbus::Struct< std::string, std::map< std::string, sdbus::Variant >>`
Is this correct?
When I call the method, I get the following error:
`terminate called after throwing an instance of 'sdbus::Error'
  what():  [System.Error.ENXIO] Failed to enter a struct (No such device or address)`
What does this error tell me and how can I fix it?

Additionally: `sdbus::Struct` is implemented as std::tuple. Can I use the standard procedure to iterate through the container?

Thanks a lot.

**Comment by sangelovic** [comment, marked as answer]
The corresponding C++ type for `a(sa{sv})` is `std::vector<sdbus::Struct<std::string, std::map<std::string, sdbus::Variant>>>`. The error tells you that are going to deserialize a struct, but there is an object of another type serialized in the D-Bus message at that location (an array in this case).

You can use the same operations on `sdbus::Struct` that you'd use on a `std::tuple`. In newer sdbus-c++ versions, the `sdbus::Struct` fully implements a tuple protocol, which makes it a tuple-like class (https://github.com/Kistler-Group/sdbus-cpp/pull/343).

`v1.0` is quite old. I suggest you upgrade to one of the latest versions (`v1.6` or, even better, `v2.0`).
